### PR TITLE
create oauth-server for bundle

### PR DESCRIPTION
**Pull request by jkralik**

### CODE REVIEW

File: authorization/provider/auth0.go [hunk 1]
```diff
@@ -54,9 +63,6 @@ func (p *Auth0Provider) LogoutURL(returnTo string) string {
 
 // Exchange Auth Code for Access Token via OAuth
 func (p *Auth0Provider) Exchange(ctx context.Context, authorizationProvider, authorizationCode string) (*Token, error) {
-	if p.GetProviderName() != authorizationProvider {
```
Comment (ondrejtomcik) [inline authorization/provider/auth0.go:57]: why did you drop verification of the apn?

File: authorization/test/service.go [hunk 2]
```diff
@@ -13,12 +13,15 @@ import (
...
+	oauthService "github.com/plgd-dev/cloud/oauth-server/service"
+	"github.com/plgd-dev/cloud/oauth-server/uri"
+	"github.com/plgd-dev/cloud/test/config"
 	testCfg "github.com/plgd-dev/cloud/test/config"
 	"github.com/plgd-dev/kit/security/certManager"
 )
 
 func newService(config service.Config, tlsConfig *tls.Config) (*service.Server, error) {
-	oauth := provider.NewTestProvider()
+	oauth := provider.NewAuth0Provider(config.Device, tlsConfig)
```
Comment (ondrejtomcik) [inline authorization/test/service.go]: why did you start to use auth0 as a default provider also for tests? (what was the motivation to drop test oauth provider)

File: bundle/README.md [hunk 3]
```diff
@@ -34,23 +34,32 @@ Image can be configured via enviroment variables as argument `-e ENV=VALUE` of c
...
-| `RESOURCE_AGGREGATE_MONGO_MAX_PARALLEL_QUERIES` | uin16 | maximum number of queries MongoDB client can execute in parallel | `8` |
-| `INITIALIZE_CERITIFICATES` | bool | initialze certificates | `true` |
-| `CERITIFICATES_PATH` | string | path to directory | `"/data/certs"` |
+| `GRPC_GATEWAY_ADDRESS`| string | secure grpc-tcp listen address | `"localhost:9084"` |
+| `HTTP_GATEWAY_ADDRESS`| string | secure grpc-tcp listen address | `"localhost:9086"` |
+| `CERTIFICATE_AUTHORITY_ADDRESS` | string | secure grpc-tcp listen address | `"localhost:9087"` |
+| `OAUTH_SERVER_ADDRESS` | string | secure grpc-tcp listen address | `"localhost:9088"` |
+| `RESOURCE_AGGREGATE_ADDRESS` | string | secure grpc-tcp listen address | `"localhost:9083"` |
+| `RESOURCE_DIRECTORY_ADDRESS` | string | secure grpc-tcp listen address | `"localhost:9082"` |
+| `INITIALIZE` | bool | initialze certificates, keys, nginx configuration | `true` |
```
Comment (ondrejtomcik) [inline bundle/README.md]: Why should user even disable initialization? Why it's configurable?

File: bundle/README.md [hunk 4]
```diff
@@ -34,23 +34,32 @@ Image can be configured via enviroment variables as argument `-e ENV=VALUE` of c
...
+| `CERTIFICATES_PATH` | string | path to directory | `"/data/certs"` |
 | `MONGO_PATH` | string | path to directory | `"/data/db"` |
 | `MONGO_PORT` | uint16 | mongo listen port  | `"10000"` |
 | `NATS_PORT` | uint16 | nats listen port  | `"10001"` |
 | `LOGS_PATH` | string | path to directory | `"/data/log"` |
+| `NGINX_PATH` | string | path to directory | `"/data/nginx"` |
```
Comment (ondrejtomcik) [inline bundle/README.md]: Is this path to nginx binaries or nginx configuration file?

File: bundle/README.md [hunk 5]
```diff
@@ -34,23 +34,32 @@ Image can be configured via enviroment variables as argument `-e ENV=VALUE` of c
...
+| `NGINX_HTTPS_PORT` | uint16 | nginx https port | `"8443"` |
```
Comment (ondrejtomcik) [inline bundle/README.md]: why this strange default port? Why not 443 as default?

File: bundle/client/grpc/main.go [hunk 6]
```diff
@@ -66,10 +68,50 @@ func decodePayload(resp *pbGW.Content) {
...
+	token := body["access_token"]
+	if token == "" {
+		return "", fmt.Errorf("token not found in body")
+	}
+	return token, nil
+}
+
 func main() {
-	addr := flag.String("addr", "localhost:9084", "address")
+	addr := flag.String("addr", "localhost:8443", "address")
```
Comment (ondrejtomcik) [inline bundle/client/grpc/main.go:112]: address localhost:8443 is configurable and reserved by the nginx; not ? or even if you set FQDN to something else this will work - because it;s inside of the container? Won't you have problems with CORS?

File: bundle/nginx/cors.conf [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    #
+    # Tell client that this pre-flight info is valid for 20 days
+    #
+    add_header 'Access-Control-Allow-Credentials' 'true';
+    add_header 'Access-Control-Max-Age' 1728000;
+    add_header 'Content-Type' 'text/plain; charset=utf-8';
+    add_header 'Content-Length' 0;
+    return 204;
+}
+if ($request_method = 'POST') {
```
Comment (ondrejtomcik) [inline bundle/nginx/cors.conf]: What about DELETE? It will be soon supported.

File: bundle/nginx/mime.types [hunk 8]
```diff
@@ -0,0 +1,135 @@
+types {
```
Comment (ondrejtomcik) [inline bundle/nginx/mime.types:1]: Did you test websockets if they are working correctly on the bundle?

File: oauth-server/cmd/service/main.go [hunk 9]
```diff
@@ -0,0 +1,22 @@
+package main
```
Comment (ondrejtomcik) [inline oauth-server/cmd/service/main.go]: Why did you introduced this as a new service on the root plgd cloud project structure (folder)? It should be part of the authorization server. It's misleading to have it there and it's just a mock for testing and bundle.

File: oauth-server/test/config/httpgw.yaml [hunk 10]
```diff
@@ -0,0 +1,69 @@
+Address: 0.0.0.0:7000
```
Comment (ondrejtomcik) [inline oauth-server/test/config/httpgw.yaml]: why is httpgw config part of the oauth server? and /home/krz/go is probably not what you wanted to commit :)

File: resource-aggregate/cqrs/eventstore/mongodb/config.go [hunk 11]
```diff
@@ -38,8 +38,8 @@ func WithTLS(cfg *tls.Config) Option {
 type Config struct {
 	URI             string        `long:"uri" env:"URI" envconfig:"URI" default:"mongodb://localhost:27017"`
 	DatabaseName    string        `long:"dbName" env:"DATABASE" envconfig:"DATABASE" default:"eventStore"`
-	BatchSize       int           `long:"batchSize" env:"BATCH_SIZE" envconfig:"BATCH_SIZE" default:"16"`
-	MaxPoolSize     uint64        `long:"maxPoolSize" env:"MAX_POOL_SIZE" envconfig:"MAX_POOL_SIZE" default:"16"`
+	BatchSize       int           `long:"batchSize" env:"BATCH_SIZE" envconfig:"BATCH_SIZE" default:"64"`
+	MaxPoolSize     uint64        `long:"maxPoolSize" env:"MAX_POOL_SIZE" envconfig:"MAX_POOL_SIZE" default:"64"`
```
Comment (ondrejtomcik) [inline resource-aggregate/cqrs/eventstore/mongodb/config.go]: Was this change intended? Why you are changing it for 4times more?

**Comment by ondrejtomcik** [comment]
There are still issues with cookies which needs to be resolved without merging. 
And UI looks strange without displaying name / email in the topbar. please fill id token properly so it's not empty and looking buggy.

File: authorization/provider/auth0.go (hunk 1 above)
Comment (jkralik) [reply authorization/provider/auth0.go:57]: It doesn't bring any security and we support only one `apn`. For user that use the bundle, it is more simple to onboard the device.

File: authorization/test/service.go (hunk 2 above)
Comment (jkralik) [reply authorization/test/service.go]: We don't need it any more. Tests uses directly oauth-server.

File: bundle/README.md (hunk 3 above)
Comment (jkralik) [reply bundle/README.md]: For example if you want to restart bundle without recreating certificates, keys, ...

File: authorization/provider/auth0.go (hunk 1 above)
Comment (ondrejtomcik) [reply authorization/provider/auth0.go:57]: this value is retrieved from the well-known cloud endpoint. User is not even aware of this value unless he is onboarding the device manually. This verification is required by the specification and we want to stay spec conformant.

File: authorization/test/service.go (hunk 2 above)
Comment (ondrejtomcik) [reply authorization/test/service.go:24]: Is this provider auth0 specific or generic?

File: resource-aggregate/cqrs/eventstore/mongodb/config.go (hunk 11 above)
Comment (jkralik) [reply resource-aggregate/cqrs/eventstore/mongodb/config.go]: No it doesn't need to be changed in future  I want to drive default value by num cores.

File: bundle/README.md (hunk 4 above)
Comment (jkralik) [reply bundle/README.md]: Removed

File: bundle/README.md (hunk 5 above)
Comment (jkralik) [reply bundle/README.md]: 443 was set

File: bundle/nginx/cors.conf (hunk 7 above)
Comment (jkralik) [reply bundle/nginx/cors.conf]: Fixed

File: bundle/nginx/mime.types (hunk 8 above)
Comment (jkralik) [reply bundle/nginx/mime.types:1]: Yes

File: oauth-server/cmd/service/main.go (hunk 9 above)
Comment (jkralik) [reply oauth-server/cmd/service/main.go]: moved to test/oauth-server

File: oauth-server/test/config/httpgw.yaml (hunk 10 above)
Comment (jkralik) [reply oauth-server/test/config/httpgw.yaml]: removed

File: bundle/nginx/nginx.conf.template [hunk 12]
```diff
@@ -0,0 +1,95 @@
+events {
```
Comment (ondrejtomcik) [inline bundle/nginx/nginx.conf.template:1]: please configure default redirect from http to https

File: authorization/provider/auth0.go (hunk 1 above)
Comment (jkralik) [reply authorization/provider/auth0.go:57]: OK - Added

File: bundle/nginx/nginx.conf.template (hunk 12 above)
Comment (jkralik) [reply bundle/nginx/nginx.conf.template:1]: I don't add it because than i need add the port to cfg.

File: bundle/README.md (hunk 3 above)
Comment (ondrejtomcik) [reply bundle/README.md]: But first time you need to initialize it, and then you don't have possibility to stop container and start again with different env variable as you specify env variable during docker run, not during docker start.

File: authorization/provider/auth0.go (hunk 1 above)
Comment (ondrejtomcik) [reply authorization/provider/auth0.go:57]: Did you change it to `plgd` please? Can you please update also cloud server in iotivity-lite? (there is pending MR with CREATE support, you can update that one).

**Comment by ondrejtomcik** [comment]
are /api/authz/code and token endpoints working ?

File: authorization/provider/auth0.go (hunk 1 above)
Comment (jkralik) [reply authorization/provider/auth0.go:57]: Updated + https://gitlab.iotivity.org/iotivity/iotivity-lite/-/merge_requests/180/diffs?commit_id=7c676b0848c72abf7318830f8ecafd7bde1b0d53

File: bundle/README.md (hunk 3 above)
Comment (jkralik) [reply bundle/README.md]: Ofcourse you can change it.

Comment (jkralik) [reply bundle/README.md]: Ok. Removed

File: bundle/client/grpc/main.go (hunk 6 above)
Comment (jkralik) [reply bundle/client/grpc/main.go:112]: grpc is not a part of bundle - you need to compile it separately

Comment (jkralik) [reply bundle/client/grpc/main.go:112]: CORS is part of browser it is not check by servers.

**Comment by jkralik** [comment]
> 
> 
> There are still issues with cookies which needs to be resolved without merging.
> And UI looks strange without displaying name / email in the topbar. please fill id token properly so it's not empty and looking buggy.

Fixed

**Review by ondrejtomcik — APPROVED** [review]
